### ISSUE TITLE
Constant also generated outside impl block

### ISSUE THREAD

**Issue by ultimaweapon** [GrayJack/bitflag-attr#1]
Hello, I'm using this crate like the following code:

```rust
/// Flags to control behavior of [`ProcMgr::fork()`].
#[bitflag(u32)]
pub enum Fork {
    /// Copy file descriptor table to the child instead of sharing it with the parent.
    CopyFd = 0x04, // RFFDG
}
```

It generate the following code:

```rust
#[repr(transparent)]
#[derive(Clone, Copy)]
#[doc = " Flags to control behavior of [`ProcMgr::fork()`]."]
pub struct Fork(u32);

#[doc = " Copy file descriptor table to the child instead of sharing it with the parent."]
#[allow(non_upper_case_globals, dead_code, unused)]
const CopyFd: u32 = 0x04;
#[allow(non_upper_case_globals)]
impl Fork {
    #[doc = " Copy file descriptor table to the child instead of sharing it with the parent."]
    pub const CopyFd: Self = Self(0x04);
    #[doc = r" Return the underlying bits of the bitflag"]
    #[inline]
    pub const fn bits(&self) -> u32 {
        self.0
    }
}
```

There are two `CopyFd` created, the first one outside the `impl` block and another one is inside. Would like to ask what is the purpose of this? Is it a bug or intended behavior? If it is intended behavior is it possible to suppress this?

Thanks!

**Comment by GrayJack** [comment]
Hey there!

So, this in the intended behaviour, to support for flags that are defined by other flags. Otherwise this crate would need to parse and handle all possible expressions that can be used in the enum syntax (which is quite a lot and recursive) and transform the same intention to the inside block constant .

What if the major thing it is bothering you, the const showing up in the docs (which is one thing that is in my backlog that I find suboptimal of this crate)? Or the const being created?

For the first, is a simple fix with `doc(hidden)`. For the second one, I guess we could have a flag or a parameter to the macro that indicates/promisses you won't create a flag based on other flags in the expression (right) side and don't generate the outside constants.

Of course, another option is to implement the parsing and handling of possible expressions. It would solve both issues, but it would be a quite good amount of work, and it would probably be a `help needed` issue.

Would either first or second option solve your concerns in the short to mid time frame?

**Comment by ultimaweapon** [comment]
Thanks for a fast response!

My concern is those constants may conflict with the other identifiers in the same scope. For example, if I want to create the following type in the same scope as `Fork`:

```rust
pub struct CopyFd {
   ...
}
```

I believe it won't be possible since it will conflict with the generated constant. Of course I can move `Fork` to a separated module but it would be great if we don't have this limitation. I'm already switched to https://crates.io/crates/bitfield-struct so I believe a proper solution that is a `help needed` should be better than an interim solution like `doc(hidden)` or a flag to prevent those constants from generated.

**Comment by GrayJack** [comment]
That is great that you found your solution! bitfield and bitflags are a different thing at the abstraction level (although they share some similarities at the low level.

Yeah, I want the complete thing at some point, it is already in my backlog, thanks for showing that there is interest from others as well!

Happy rust hacking!

**Comment by GrayJack** [comment]
Fixed by 3d9a0548a6ea62cee8fd28ef7b767254dc561b89
